extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/discovery/MultiConnectionTopicSubscribeService.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.binding.mqtt.discovery;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ *
+ * The {@link MultiConnectionTopicSubscribeService} is responsible for subscribing to a topic on all
+ * available broker connections.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public interface MultiConnectionTopicSubscribeService {
```

Comment:
Can you tune the naming, so it feels more like an `MQTTMultiBrokerDiscovery` and the TopicMessageReceived as an `MQTTDiscoveryParticipant`?
Then `MQTTDiscoveryParticipant` should also define a method `getDiscoveryTopic`, so for this interface's `subscribe` method only an instance of the participant is needed.

It would be more close the the current DiscoveryService <-> DiscoveryParticipant pattern we have in other places. The functionality is already what we want, so its all about the names.

Reply:
And another thing: right now the BrokerHandlerFactory itself is this discovery service. What about having a dedicated service being registered from the BrokerHandlerFactory and feed with new BrokerHandlers. It would better separate the concerns. wdyt?

Reply:
I have tried that first, but that service would not do much actually. The service implementation added around 50 lines more to the factory class file. In my opinion there is no further separation required.

Reply:
I'm not sure about the naming yet. The current one sound a bit clumsy, I agree. The service is not for detecting multiple brokers though, but for subscribing to a topic on all available (and future) brokers connections.
Maybe something around "MQTTMultiBrokerSubscribe" and "MQTTSubscribeParticipant".

Reply:
The main thing for me is that we talk about "discovery". So having DiscoveryService and DiscoveryParticipant would match our current terminology.